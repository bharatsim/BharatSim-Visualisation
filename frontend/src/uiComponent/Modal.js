import React from 'react';
import PropTypes from 'prop-types';

import { withStyles } from '@material-ui/core/styles';
import Button from '@material-ui/core/Button';
import Dialog from '@material-ui/core/Dialog';
import MuiDialogTitle from '@material-ui/core/DialogTitle';
import MuiDialogContent from '@material-ui/core/DialogContent';
import MuiDialogActions from '@material-ui/core/DialogActions';
import IconButton from '@material-ui/core/IconButton';
import CloseIcon from '@material-ui/icons/Close';
import Typography from '@material-ui/core/Typography';
import Box from '@material-ui/core/Box';
import ButtonGroup from './ButtonGroup';

const styles = (theme) => ({
  root: {
    margin: 0,
    padding: theme.spacing(4, 5, 4, 6),
  },
  // Todo: remove this stying added at global level
  closeButton: {
    height: theme.spacing(8),
    width: theme.spacing(8),
    color: theme.palette.grey['800'],
  },
  titleContainer: {
    display: 'flex',
    justifyContent: 'space-between',
    alignItems: 'center',
  },
});

const DialogContent = withStyles((theme) => ({
  root: {
    margin: theme.spacing(0, 4),
    padding: theme.spacing(4, 0),
  },
}))(MuiDialogContent);

const DialogActions = withStyles((theme) => ({
  root: {
    margin: 0,
    padding: theme.spacing(4),
  },
}))(MuiDialogActions);

const DialogTitle = withStyles(styles)((props) => {
  const { children, classes, onClose, ...other } = props;
  return (
    <MuiDialogTitle disableTypography className={classes.root} {...other}>
      <Box className={classes.titleContainer}>
        <Typography variant="h5">{children}</Typography>
        <IconButton
          aria-label="close"
          data-testid="button-icon-close"
          className={classes.closeButton}
          onClick={onClose}
        >
          <CloseIcon />
        </IconButton>
      </Box>
    </MuiDialogTitle>
  );
});

function Modal({ open, handleClose, title, children, actions }) {
  return (
    <Dialog onClose={handleClose} aria-labelledby="customized-dialog-title" open={open}>
      <DialogTitle id="customized-dialog-title" onClose={handleClose}>
        {title}
      </DialogTitle>
      <DialogContent dividers>{children}</DialogContent>
      {actions.length !== 0 && (
        <DialogActions>
          <ButtonGroup>
            {actions.map(({ name, handleClick, variant, isDisable, color }) => (
              <Button
                key={name}
                data-testid={`button-${name}`}
                autoFocus
                onClick={handleClick}
                variant={variant}
                color={color}
                disabled={isDisable}
              >
                {name}
              </Button>
            ))}
          </ButtonGroup>
        </DialogActions>
      )}
    </Dialog>
  );
}

Modal.defaultProps = {
  actions: [],
};

Modal.propTypes = {
  open: PropTypes.bool.isRequired,
  handleClose: PropTypes.func.isRequired,
  title: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([PropTypes.element, PropTypes.string]).isRequired,
  actions: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string.isRequired,
      handleClick: PropTypes.func.isRequired,
      variant: PropTypes.string.isRequired,
      color: PropTypes.string,
      isDisable: PropTypes.bool,
    }),
  ),
};

export default Modal;
export { DialogActions, DialogContent, DialogTitle };
