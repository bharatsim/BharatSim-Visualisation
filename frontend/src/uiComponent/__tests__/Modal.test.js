import React from 'react';
import { render, within } from '@testing-library/react';

import { fireEvent } from '@testing-library/dom';
import Modal from '../Modal';

describe('<Modal />', () => {
  it('should match snapshot', () => {
    render(
      <Modal
        open
        handleClose={jest.fn()}
        title="Modal"
        actions={[
          { name: 'Apply', handleClick: jest.fn(), variant: 'contained', color: 'primary' },
        ]}
      >
        Hello this is modal
      </Modal>,
    );

    expect(document.querySelector('.MuiPaper-root')).toMatchSnapshot();
  });

  it('should match snapshot without actions', () => {
    render(
      <Modal open handleClose={jest.fn()} title="Modal">
        Hello this is modal
      </Modal>,
    );

    expect(document.querySelector('.MuiPaper-root')).toMatchSnapshot();
  });

  it('should call handle close callback on click of close icon button', () => {
    const handleClose = jest.fn();
    render(
      <Modal
        open
        handleClose={handleClose}
        title="Modal"
        actions={[
          { name: 'Apply', handleClick: jest.fn(), variant: 'contained', color: 'primary' },
        ]}
      >
        Hello this is modal
      </Modal>,
    );

    const modal = within(document.querySelector('.MuiPaper-root'));
    const closeIconButton = modal.getByTestId('button-icon-close');

    fireEvent.click(closeIconButton);

    expect(handleClose).toHaveBeenCalled();
  });

  it('should call action callback on click of first action button', () => {
    const handleApply = jest.fn();
    render(
      <Modal
        open
        handleClose={jest.fn()}
        title="Modal"
        actions={[
          { name: 'Apply', handleClick: handleApply, variant: 'contained', color: 'primary' },
        ]}
      >
        Hello this is modal
      </Modal>,
    );

    const modal = within(document.querySelector('.MuiPaper-root'));
    const closeIconButton = modal.getByTestId('button-Apply');

    fireEvent.click(closeIconButton);

    expect(handleApply).toHaveBeenCalled();
  });
});
