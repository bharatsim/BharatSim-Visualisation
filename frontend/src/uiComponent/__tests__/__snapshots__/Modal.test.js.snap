// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should match snapshot 1`] = `
<div
  aria-labelledby="customized-dialog-title"
  class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
  role="dialog"
>
  <div
    class="MuiDialogTitle-root Component-root-1"
    id="customized-dialog-title"
  >
    <div
      class="MuiBox-root MuiBox-root-4 Component-titleContainer-3"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5"
      >
        Modal
      </h5>
      <button
        aria-label="close"
        class="MuiButtonBase-root MuiIconButton-root Component-closeButton-2"
        data-testid="button-icon-close"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiDialogContent-root WithStyles(ForwardRef(DialogContent))-root-5 MuiDialogContent-dividers"
  >
    Hello this is modal
  </div>
  <div
    class="MuiDialogActions-root WithStyles(ForwardRef(DialogActions))-root-6 MuiDialogActions-spacing"
  >
    <div
      class="MuiBox-root MuiBox-root-8 makeStyles-buttonGroup-7"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary Mui-focusVisible Mui-focusVisible"
        data-testid="button-Apply"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiButton-label"
        >
          Apply
        </span>
        <span
          class="MuiTouchRipple-root"
        >
          <span
            class="MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
            style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
          >
            <span
              class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
            />
          </span>
        </span>
      </button>
    </div>
  </div>
</div>
`;

exports[`<Modal /> should match snapshot without actions 1`] = `
<div
  aria-labelledby="customized-dialog-title"
  class="MuiPaper-root MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthSm MuiPaper-elevation24 MuiPaper-rounded"
  role="dialog"
>
  <div
    class="MuiDialogTitle-root Component-root-9"
    id="customized-dialog-title"
  >
    <div
      class="MuiBox-root MuiBox-root-12 Component-titleContainer-11"
    >
      <h5
        class="MuiTypography-root MuiTypography-h5"
      >
        Modal
      </h5>
      <button
        aria-label="close"
        class="MuiButtonBase-root MuiIconButton-root Component-closeButton-10"
        data-testid="button-icon-close"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiDialogContent-root WithStyles(ForwardRef(DialogContent))-root-13 MuiDialogContent-dividers"
  >
    Hello this is modal
  </div>
</div>
`;
